firebase-perf/src/test/java/com/google/firebase/perf/application/FrameMetricsRecorderTest.java

```diff
@@ -0,0 +1,405 @@
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.ArgumentMatchers.nullable;
+import static org.mockito.Mockito.doCallRealMethod;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+import android.view.WindowManager;
+import androidx.core.app.FrameMetricsAggregator;
+import com.google.firebase.perf.FirebasePerformanceTestBase;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Optional;
+import java.util.List;
+import java.util.WeakHashMap;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Spy;
+import org.robolectric.Robolectric;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.android.controller.ActivityController;
+
+/** Unit tests for {@link com.google.firebase.perf.application.FrameMetricsRecorder}. */
+@RunWith(RobolectricTestRunner.class)
+public class FrameMetricsRecorderTest extends FirebasePerformanceTestBase {
+  private Activity activity;
+  private FrameMetricsRecorder recorder;
+
+  @Mock private FrameMetricsAggregator fma;
+  @Mock private PerfFrameMetrics frameMetrics1;
+  @Mock private PerfFrameMetrics frameMetrics2;
+  @Mock private PerfFrameMetrics frameMetrics3;
+
+  @Spy private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap = new WeakHashMap<>();
+
+  @Before
+  public void setUp() {
+    initMocks(this);
+    activity = createFakeActivity(true);
+    recorder = spy(new FrameMetricsRecorder(activity, fma, subTraceMap));
+  }
+
+  @Test
+  public void eachInstance_usesOneAndTheSameActivityInFrameMetricsAggregator_always() {
+    stubSnapshotToDoNothing();
+    ArgumentCaptor<Activity> activityCaptor = ArgumentCaptor.forClass(Activity.class);
+    recorder.start();
+    recorder.stop();
+    recorder.start();
+    recorder.stop();
+    verify(fma, times(2)).add(activityCaptor.capture());
+    verify(fma, times(2)).remove(activityCaptor.capture());
+    List<Activity> activities = activityCaptor.getAllValues();
+    Assert.assertTrue(activities.stream().allMatch(a -> a == activity));
+  }
+
+  @Test
+  public void start_callsFrameMetricsAggregator() {
+    verify(fma, times(0)).add(nullable(Activity.class));
+    recorder.start();
+    verify(fma, times(1)).add(nullable(Activity.class));
+  }
+
+  @Test
+  public void start_whileAlreadyStarted_fails() {
+    recorder.start();
+    recorder.start();
+    verify(fma, times(1)).add(nullable(Activity.class));
```

Comment:
Do we want this to be nullable? Why are we using this argument matcher instead of a different one?